cirq/ops/linear_combinations.py

```diff
@@ -315,6 +316,90 @@ def copy(self) -> 'PauliSum':
         factory = type(self)
         return factory(self._linear_dict.copy())
 
+    def expectation_from_wavefunction(
+            self,
+            state: np.ndarray,
+            qubit_map: Dict[raw_types.Qid, int]) -> float:
+        """Evaluate the expectation of this PauliSum given a wavefunction.
+
+        See `PauliString.expectation_from_wavefunction`.
+        """
+        if any([abs(p.coefficient.imag) > 0.0001 for p in self]):
+            raise NotImplementedError(
+                "Cannot compute expectation value of a non-Hermitian "
+                "PauliString <{}>. Coefficient must be real.".format(self))
+
+        if not isinstance(qubit_map, dict) or not all(
+            isinstance(k, raw_types.Qid) and isinstance(v, int) for k, v in qubit_map.items()):
+            raise TypeError("Input qubit map must be a valid mapping from "
+                            "Qubit ID's to integer indices.")
+
+        if set(qubit_map.keys()) < set(self.qubits):
+            raise ValueError("`qubit_map` must be a complete mapping over this "
+                             "PauliString's qubits.")
+
+        # FIXME: Avoid enforce specific complex type. This is necessary to
+        # prevent an `apply_unitary` bug (Issue #2041).
+        if state.dtype.kind != 'c':
+            raise TypeError("Input state dtype must be np.complex64 or "
+                            "np.complex128")
+
+        size = state.size
+        num_qubits = size.bit_length() - 1
+        if len(state.shape) == 1 or state.shape == (2,) * num_qubits:
+            # HACK: avoid circular import
+            from cirq.sim.wave_function import validate_normalized_state
+            validate_normalized_state(state, num_qubits, dtype=state.dtype)
+            return sum([p._expectation_from_wavefunction(state, qubit_map) for p in self])
+
+        raise ValueError("Input array does not represent a wavefunction with "
+                         "shape `(2 ** n,)` or `(2, 2, ..., 2)`.")
+
+    def expectation_from_density_matrix(self,
+                    state: np.ndarray,
+                    qubit_map: Optional[Dict[raw_types.Qid, int]]=None
+                    ) -> float:
+        """Evaluate the expectation of this PauliSum given a density matrix.
+
+        See `PauliString.expectation_from_density_matrix`.
+        """
+        if any([abs(p.coefficient.imag) > 0.0001 for p in self]):
+            raise NotImplementedError(
+                "Cannot compute expectation value of a non-Hermitian "
+                "PauliString <{}>. Coefficient must be real.".format(self))
+
+        if not isinstance(qubit_map, dict) or not all(
+            isinstance(k, raw_types.Qid) and isinstance(v, int) for k, v in qubit_map.items()):
+            raise TypeError("Input qubit map must be a valid mapping from "
+                            "Qubit ID's to integer indices.")
+
+        if set(qubit_map.keys()) < set(self.qubits):
```

Comment:
The body of this method looks *extremely* similar to the wavefunction one. Can you extra a common private method that is parameterized by an argument of `sub=PauliString.expectation_from_density_matrix` or `sub=PauliString.expectation_from_wavefunction`?

Reply:
the issue with that is that Paulisum is making calls directly to the PauliString methods that have no input validation so that `state` only ever gets validated once (in the `PauliSum.expectation_from_*` method). So the `sum(p._expectation_from_*(state, qubit_map) for p in self)` return value has to be preceded by state-specific input validation that would be nastier to parametrize over.

The best that a common method could do (after adding a `_validate_qubit_mapping` function) is to provide Hermiticity and dtype validation which isn't worth an additional function def imo